### PR TITLE
System extension table in macOS

### PR DESCRIPTION
**Pull request by kumarak**
The changes add **System Extension table** for macOS which is the new way to develop app extensions with security and networking related capabilities. 

It uses the system extension database `/Library/SystemExtensions/db.plist` to get the list of extensions installed and the associated app they are getting referenced from. This is initial effort to get the extensions table and it will great to get feedback from the community about both: on the table schema and the way it should be implemented. 

The table schema includes the information from the `db.plist`. I am looking to include extension policies as well but there are not many references. It will be great if someone can help me with pointers. Also, parsing `db.plist` may not be the best way to get the list, but there are no open API's to get this. I might be missing here and it will be great if someone could provide me references. 

The PR is in reference to the feature request #6738 . The example uses of the system extension table is below:

```
osquery> select * from system_extensions;
+-----------------------------------------------------------------------------------------------------------+--------------------------------------+-------------------+----------------------------------+---------+----------------------------------------------+------------------------+------------+
| path                                                                                                      | uuid                                 | state             | identifier                       | version | category                                     | bundle_path            | team       |
+-----------------------------------------------------------------------------------------------------------+--------------------------------------+-------------------+----------------------------------+---------+----------------------------------------------+------------------------+------------+
| /Applications/LuLu.app/Contents/Library/SystemExtensions/com.objective-see.lulu.extension.systemextension | B4A2DE3D-1047-4109-9878-CC8238F6DE29 | activated_enabled | com.objective-see.lulu.extension | 2.0.0   | com.apple.system_extension.network_extension | /Applications/LuLu.app | VBG97UB4TA |
+-----------------------------------------------------------------------------------------------------------+--------------------------------------+-------------------+----------------------------------+---------+----------------------------------------------+------------------------+------------+
```

### CODE REVIEW

File: specs/darwin/system_extensions.table [hunk 1]
```diff
@@ -0,0 +1,13 @@
+table_name("system_extensions")
+description("OS X's system extension table.")
```
Comment (mike-myers-tob) [inline specs/darwin/system_extensions.table:2]: Can we use `macOS` here?

File: specs/darwin/system_extensions.table [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+schema([
+    Column("path", TEXT, "Original path of system extension"),
+    Column("uuid", TEXT, "Extension unique id"),
```
Comment (mike-myers-tob) [inline specs/darwin/system_extensions.table:5]: Nitpick: `ID` (caps) to stay consistent

File: specs/darwin/system_extensions.table (hunk 1 above)
Comment (mike-myers-tob) [reply specs/darwin/system_extensions.table:2]: Also it could say `macOS 10.15 and higher`

File: tests/integration/tables/system_extensions.cpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+class SystemExtension : public testing::Test {
+ protected:
+  void SetUp() override {
+    setUpEnvironment();
+  }
+};
+
+TEST_F(SystemExtension, test_sanity) {
+  QueryData data = execute_query("select * from system_extensions");
```
Comment (mike-myers-tob) [inline tests/integration/tables/system_extensions.cpp:26]: Test might fail on macOS 10.14, I have a VM, let me know if you want me to test there

File: osquery/tables/system/tests/darwin/system_extensions_test.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        "\"0\", \"extensionPolicies\": \"\", \"bootUUID\": "
+        "\"E7C066D9-19F4-4E47-8E1E-35E1C1434905\" }";
+
+  try {
+    pt::read_json(ss, ptree);
+  } catch (std::exception& e) {
+    LOG(ERROR) << "Failed to read plist JSON string " << e.what();
+  }
+
+  QueryData results = genExtensionsFromPtree(ptree);
```
Comment (mike-myers-tob) [inline osquery/tables/system/tests/darwin/system_extensions_test.cpp:82]: You could have a 
`ASSERT_EQ(results.size(), 1U);`
here

File: osquery/tables/system/tests/darwin/system_extensions_test.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+namespace tables {
+
+extern QueryData genExtensionsFromPtree(pt::ptree& ptree);
+
+class SystemExtensionTests : public testing::Test {};
+
+TEST_F(SystemExtensionTests, test_parse_plist) {
+  pt::ptree ptree;
+  std::stringstream ss;
+  ss << "{ \"version\": \"1\", \"extensions\": [{ \"state\": "
```
Comment (mike-myers-tob) [inline osquery/tables/system/tests/darwin/system_extensions_test.cpp:31]: Some of the other tests store the mock file as a file in the test config directory.
E.g., `auto dbPlist_path = getTestConfigDirectory() / "db.plist";`

Which currently is here: https://github.com/osquery/osquery/tree/master/tools/tests/configs

File: osquery/tables/system/darwin/system_extensions.mm [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+void getExtensionRow(const pt::ptree& extension, Row& r) {
+  r["path"] = extension.get("originPath", "");
+  r["uuid"] = extension.get("uniqueID", "");
+  r["state"] = extension.get("state", "");
+  r["identifier"] = extension.get("identifier", "");
+  r["version"] = extension.get("bundleVersion.CFBundleShortVersionString", "");
+  r["team"] = extension.get("teamID", "");
+
+  // Get the system extension catagories from the array
```
Comment (mike-myers-tob) [inline osquery/tables/system/darwin/system_extensions.mm:47]: Typo: `categories`

**Review by mike-myers-tob — COMMENTED** [review]
Tested working on macOS 11.0.1
Nice work :+1:

File: tests/integration/tables/system_extensions.cpp (hunk 3 above)
Comment (mike-myers-tob) [reply tests/integration/tables/system_extensions.cpp:26]: > Test might fail on macOS 10.14, I have a VM, let me know if you want me to test there

Disregard this. It built and tested fine on macOS 10.14.

**Review by theopolis — CHANGES_REQUESTED** [review]
Setting this to 'request changes' for Mike's feedback.

File: osquery/tables/system/darwin/system_extensions.mm [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+  r["version"] = extension.get("bundleVersion.CFBundleShortVersionString", "");
+  r["team"] = extension.get("teamID", "");
+
+  // Get the system extension categories from the array
+  const auto category = extension.get_child("categories");
+  r["category"] = getExtensionCategory(category);
+  r["bundle_path"] = extension.get("container.bundlePath", "");
+}
+
+QueryData genExtensionsFromPtree(pt::ptree& ptree) {
```
Comment (theopolis) [inline osquery/tables/system/darwin/system_extensions.mm:53]: Can this be `const pt::ptree&`?

File: osquery/tables/system/darwin/system_extensions.mm [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    Row r;
+    getExtensionRow(extension_value, r);
+    results.push_back(r);
+  }
+  return results;
+}
+
+QueryData genSystemExtensions(QueryContext& context) {
+  if (!osquery::pathExists(kSysExtDBPath)) {
+    VLOG(1) << "System extension database doesn't exist";
```
Comment (theopolis) [inline osquery/tables/system/darwin/system_extensions.mm:72]:
```suggestion
    VLOG(1) << "System extension database does not exist: " << kSysExtDBPath;
```

File: osquery/tables/system/darwin/system_extensions.mm [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    return {};
+  }
+
+  pt::ptree ptree;
+  if (!osquery::parsePlist(kSysExtDBPath, ptree).ok()) {
+    LOG(ERROR) << "Failed to parse " << kSysExtDBPath;
```
Comment (theopolis) [inline osquery/tables/system/darwin/system_extensions.mm:78]:
```suggestion
    LOG(ERROR) << "Failed to parse: " << kSysExtDBPath;
```

File: osquery/tables/system/tests/darwin/system_extensions_test.cpp [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+#include <boost/algorithm/string.hpp>
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/property_tree/ptree.hpp>
+#include <gtest/gtest.h>
+
+#include <osquery/config/tests/test_utils.h>
+#include <osquery/core/sql/query_data.h>
+#include <osquery/filesystem/filesystem.h>
+#include <osquery/logger/logger.h>
+#include <osquery/utils/darwin/plist.h>
```
Comment (theopolis) [inline osquery/tables/system/tests/darwin/system_extensions_test.cpp:19]: Do we need all of these?

**Review by theopolis — CHANGES_REQUESTED** [review]
Some quick nitpicks then looks good!

File: osquery/tables/system/tests/darwin/system_extensions_test.cpp (hunk 10 above)
Comment (kumarak) [reply osquery/tables/system/tests/darwin/system_extensions_test.cpp]: Thanks, @theopolis for the review. I cleaned up and removed some of the headers which are indirectly getting included.

**Review by theopolis — APPROVED** [review]

**Pull request merged**